src/memory.rs

```diff
@@ -0,0 +1,153 @@
+}
+
+impl BinaryMultiple {
+    pub fn to_legacy(&self) -> String {
+        match self {
+            BinaryMultiple::Kibi => "k".to_string(),
+            BinaryMultiple::Mebi => "m".to_string(),
+            BinaryMultiple::Gibi => "g".to_string(),
+            BinaryMultiple::Tebi => "t".to_string(),
+            BinaryMultiple::Pebi => "p".to_string(),
+            BinaryMultiple::Exbi => "e".to_string(),
+        }
+    }
+
+    pub fn upscale(&self) -> Self {
+        match self {
+            BinaryMultiple::Kibi => BinaryMultiple::Kibi,
+            BinaryMultiple::Mebi => BinaryMultiple::Kibi,
+            BinaryMultiple::Gibi => BinaryMultiple::Mebi,
+            BinaryMultiple::Tebi => BinaryMultiple::Gibi,
+            BinaryMultiple::Pebi => BinaryMultiple::Tebi,
+            BinaryMultiple::Exbi => BinaryMultiple::Pebi,
+        }
+    }
+}
+
+impl FromStr for BinaryMultiple {
+    type Err = Error;
+
+    fn from_str(q: &str) -> OperatorResult<BinaryMultiple> {
+        let lq = q.to_lowercase();
+        match lq.as_str() {
+            "ki" | "kib" => Ok(BinaryMultiple::Kibi),
+            "mi" | "mib" => Ok(BinaryMultiple::Mebi),
+            "gi" | "gib" => Ok(BinaryMultiple::Gibi),
+            "ti" | "tib" => Ok(BinaryMultiple::Tebi),
+            "pi" | "pib" => Ok(BinaryMultiple::Pebi),
+            "ei" | "eib" => Ok(BinaryMultiple::Exbi),
+            _ => Err(Error::InvalidQuantityUnit {
+                value: q.to_string(),
+            }),
+        }
+    }
+}
+
+/// Easily transform K8S memory resources to Java heap options.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Memory {
+    value: f32,
+    unit: BinaryMultiple,
+}
+
+impl Memory {
+    /// Scale by the given factor. If the factor is less then one
+    /// the unit granularity is increased one level to ensure eventual
+    /// conversions to Java heap settings don't end up with zero values..
+    pub fn scale(&self, factor: f32) -> Self {
+        if factor < 1.0 && self.unit != BinaryMultiple::Kibi {
+            Memory {
+                value: self.value * factor * 1024.0,
+                unit: self.unit.upscale(),
+            }
+        } else {
+            Memory {
+                value: self.value * factor,
+                unit: self.unit.clone(),
+            }
+        }
+    }
+
+    /// The Java heap settings do not support fractional values therefore
+    /// this cannot be implemented without loss of precision.
+    pub fn to_java_heap(&self, factor: f32) -> String {
+        let scaled = self.scale(factor);
+        format!("-Xmx{:.0}{}", scaled.value, scaled.unit.to_legacy())
+    }
+}
+
+impl FromStr for Memory {
+    type Err = Error;
+
+    fn from_str(q: &str) -> OperatorResult<Self> {
+        let mut v = String::from("");
+        let mut u = String::from("");
+
+        for c in q.chars() {
+            if c.is_numeric() || c == '.' {
+                v.push(c);
+            } else {
+                u.push(c);
+            }
```

Comment:
1. This will allow you to continue reading numbers after the unit
2. No need to copy the quantity string here, you can just slice the `&str`.. for example:

```rust
let start_of_unit = q.find(|c| c != '.' && !c.is_numeric()).context(NoUnitSnafu)?;
let (value, unit) = q.split_at(start_of_unit);
```